subsys/ipc/ipc_service/lib/icmsg.c

```diff
@@ -469,12 +487,21 @@ int icmsg_release_rx_buffer(const struct icmsg_config_t *conf,
 		return -EINVAL;
 	}
 
-	was_held = atomic_cas(&dev_data->rx_buffer_held,
-			      RX_BUFFER_HELD, RX_BUFFER_RELEASED);
+	/* Do not schedule new packet processing until buffer will be released.
+	 * Protect buffer against being freed multiple times.
+	 */
+	was_held = atomic_cas(&dev_data->rx_buffer_state,
+			      RX_BUFFER_STATE_HELD, RX_BUFFER_STATE_RELEASING);
 	if (!was_held) {
 		return -EALREADY;
 	}
 
+	spsc_pbuf_free(dev_data->rx_ib, dev_data->rx_len);
+	dev_data->rx_buffer = NULL;
+	dev_data->rx_len = 0;
+
+	atomic_set(&dev_data->rx_buffer_state, RX_BUFFER_STATE_RELEASED);
```

Comment:
The atomic_x operations can be implemented differently for various architectures. Mostly it is generic Zephyr implementation where spinlock is used so basically it is a irq lock/unlcok under. But the issue might be there because for ARM V7/V8 M there is synchronization instruction to ensure that lock or unlock is effective immediately, for RISCV it looks like we don't use barrier so there might be not guaranty that irq_lock/unlock will be executed in requested order. Anyway I am not an expert in our VPRs architecture. I think this possible issue should be solved under atomic implementation.

Reply:
Fully agree it should be solved in atomics API and their implementations. Not in this PR